lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,398 @@
+//! needs.
+//!
+//! Example usage
+//! --------------
+//! ```
+//! let some_task = || "DoWork";
+//!
+//! // Manually start and end the measurement.
+//! let gui_init = profiling::task::start("GUI initialization");
+//! some_task();
+//! gui_init.end();
+//! // Or use the `measure` method.
+//! profiling::task::measure("GUI initialization", some_task);
+//! ```
+//!
+//! Note that this API and encoding formats for messages are synced with the JS equivalent in
+//! `app/ide-desktop/lib/profiling/src/profiling.ts`.
+#![warn(missing_debug_implementations)]
+#![warn(missing_docs)]
+#![warn(trivial_casts)]
+#![warn(trivial_numeric_casts)]
+#![warn(unsafe_code)]
+#![warn(unused_import_braces)]
+
+use enso_prelude::*;
+
+use enso_prelude::fmt::Formatter;
+use enso_web::performance;
+use inflector::Inflector;
+use ordered_float::OrderedFloat;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_plain::from_str;
+use web_sys::PerformanceEntry;
+
+
+
+// =================
+// === Log Level ===
+// =================
+
+#[derive(Copy, Clone, Debug)]
+#[derive(Serialize, Deserialize)]
+#[allow(missing_docs)]
+pub enum ProfilingLevel {
+    Section,
+    Task,
+    Detail,
+    Debug,
+}
+
+
+// ===============
+// === Loggers ===
+// ===============
+
+/// Define a new boolean variable whose value is determined by a feature flag in the crate.
+/// The name of the variable is `ENABLED` and it will be true if a feature flag
+/// `enable-<log_level_name>-profiling` is set, otherwise it will be false.
+macro_rules! define_profiling_toggle {
+    ($log_level_name:ident) => {
+        paste::item! {
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(feature = "enable-" $log_level_name "-profiling")]
+            pub const ENABLED: bool = true;
+            #[doc = "Defines whether the log level `" $log_level_name "` should be used."]
+            #[cfg(not(feature = "enable-" $log_level_name  "-profiling"))]
+            pub const ENABLED: bool = false;
+        }
+    };
+}
+
+/// Define a new profiling module that exposes `start`, `end` and `measure` methods. The profiling
+/// can be activated and de-activated via a crate feature flag named
+/// `enable-<profiling_module_name>-profiling`, which will turn the profiling methods into no-ops.
+macro_rules! define_logger {
+    ($log_level:expr, $log_level_name_upper:ident, $log_level_name:ident) => {
+        /// Profiler module that exposes methods to measure named intervals.
+        pub mod $log_level_name {
+            use super::*;
+
+            define_profiling_toggle!($log_level_name);
+
+            /// Start measuring a named time interval. Return an `IntervalHandle` that can be sued
+            /// to end the profiling.
+            pub fn start(interval_name: &str) -> IntervalHandle {
+                start_interval($log_level, interval_name)
+            }
```

Comment:
Could you please also register file name and line number in these macros (as in the screenshots in the implementation doc), please? Rust allows you to just use the file!(), line!() macros.

Reply:
Done.